sapicore/data/__init__.py

```diff
@@ -0,0 +1,474 @@
+                        self.descriptors[c] = AxisDescriptor(name=c, labels=table[c].tolist())
+
+        # add descriptors from extra unnamed arguments.
+        for arg in args:
+            self.descriptors[arg.name] = arg
+
+        # add descriptors from extra AxisDescriptor arguments.
+        for key, value in kwargs.items():
+            self.descriptors[key] = value
+
+    def aggregate_descriptors(self, axis: int = 0) -> pd.DataFrame:
+        """Condenses the referenced :class:`~Data` set's :class:`~AxisDescriptor` objects into a
+        pandas dataframe `table`. Performed for row (sample) descriptors by default, but can also be
+        used to aggregate :class:`AxisDescriptor` objects describing any other `axis` of this dataset.
+
+        """
+        df = pd.DataFrame()
+        for key, values in self.descriptors.items():
+            if values.axis == axis:
+                df[key] = values.labels
+
+        return df
+
+    def load(self, indices: tuple[int] | list[tuple[int]] = None, kind: str = None):
+        """Populates the `samples` tensor buffer and/or `descriptors` attribute table by loading one or more files
+        into memory, potentially selecting only `indices`. Since different datasets have different formats,
+        users will have to override this method.
+
+        Parameters
+        ----------
+        indices: tuple of int or list of tuple of int
+            Specific indices to include, one for each file.
+
+        kind: str
+            Kind of content to load: "data" or "descriptors"/"labels" (defaults to both).
+
+        Note
+        ----
+        When it comes to data I/O, each dataset will have be treated differently, depending on size and
+        other considerations. It is up to the user to decide what is best for their particular dataset.
+
+        Warning
+        -------
+        Populating the `samples` buffer with the entire dataset should only be done when it can fit in memory.
+        For large sets, the buffer should not be used directly or naively; :meth:`access_data` should be overriden
+        to implement some form of lazy loading.
+
+        """
+        pass
+
+    def save(self, destination: str, kind: str = None):
+        """Dump the contents of the `samples` buffer and/or `descriptors` attribute table to disk at `destination`.
+        Since different datasets have different formats, users may wish to override this method.
+
+        While the default implementation saves the tensor as a .pt file, users are encouraged to override
+        this method for every dataset and pick/design a format appropriate for their data.
+
+        Parameters
+        ----------
+        destination: str
+            Path to the desired output file.
+
+        kind: str
+            Kind of content to save: "data" or "descriptors"/"labels" (saves everything by default).
+
+        """
+        if kind is None or kind == "descriptors" or kind == "labels":
```

Comment:
When I was sandboxing over winter break, I had a `Descriptor` class that owned multiple `AxisDescriptor` objects and operated on them (that can include aggregation). Then, a Data object owned a `Descriptor` instead of multiple singleton axis descriptors. I think that's where you're taking this; it made sense to me then and still does. It'll take some work but I lean towards doing it.